### ISSUE TITLE
Possible network connection leak

### ISSUE THREAD

**Issue by nlacey** [olivere/elastic#1192]
Please use the following questions as a guideline to help me answer
your issue/question without further inquiry. Thank you.

### Which version of Elastic are you using?

[ ] elastic.v7 (for Elasticsearch 7.x)
[x] elastic.v6 (for Elasticsearch 6.x)
[ ] elastic.v5 (for Elasticsearch 5.x)
[ ] elastic.v3 (for Elasticsearch 2.x)
[ ] elastic.v2 (for Elasticsearch 1.x)

### Please describe the expected behavior

Elasticsearch shouldn't have a network connection leak

### Please describe the actual behavior

On one of our nodes, we are seeing more than a million open network connections to elasticsearch

### Any steps to reproduce the behavior?

I believe our requests where being returned in a error state, (not sure the cause).
But when my IT guy said we have a million open network connections to elasticsearch.
I started looking for possible causes the the leak.
I believe I found it.
https://github.com/olivere/elastic/blob/release-branch.v7/client.go#L958
```
       res, err := c.c.Do((*http.Request)(req).WithContext(ctx))
	if err != nil {
		return nodes
	}
	if res == nil {
		return nodes
	}

	if res.Body != nil {
		defer res.Body.Close()
	}
```

This code has the same bug that I discovered a couple years ago in my own applications.
Basically it's possible for the network call Do()  to return err != nil, and res.Body != nil
It seemed to be rare, (many of the exit conditions in Do() ) return a nil Body, but I saw at least on case where it can return an error and a Body.

This code and any other code like it needs to be re-written at
```
       res, err := c.c.Do((*http.Request)(req).WithContext(ctx))
       if res != nil && res.Body != nil {
		defer res.Body.Close()
	}
	if err != nil {
		return nodes
	}
	if res == nil {
		return nodes
	}	
```

**Comment by olivere** [comment]
Oh well. Thanks for finding this. I’m not an AWS user, so it probably was dangerous to add signer and hope for the best. I will probably get to it on Friday and cut a new release by then.

**Comment by nlacey** [comment]
great.

https://github.com/olivere/elastic/pull/1193
Fixes this in all the places I could find it
